## EDA

In [54]:
# Importing libraries
import pandas as pd

In [55]:
# Read CSV files from data folder and store in dataframes
hackathons = pd.read_csv('/Users/Isha/Desktop/TUe/Year 3/BEP/BEP_project/data/hackathons.csv')
participants = pd.read_csv('/Users/Isha/Desktop/TUe/Year 3/BEP/BEP_project/data/participants.csv')
participants_links = pd.read_csv('/Users/Isha/Desktop/TUe/Year 3/BEP/BEP_project/data/participants-links.csv')
projects = pd.read_csv('/Users/Isha/Desktop/TUe/Year 3/BEP/BEP_project/data/projects.csv')

/var/folders/_f/7jgfw7qx5jn8xl64g_9px9g00000gp/T/ipykernel_8962/3353905109.py:3: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  participants = pd.read_csv('/Users/Isha/Desktop/TUe/Year 3/BEP/BEP_project/data/participants.csv')


In [56]:
print('Hackathon dataframe columns: ', hackathons.columns)
hackathons

Hackathon dataframe columns:  Index(['Unnamed: 0', 'URL', 'Criteria', 'schedule', 'hack_type', 'info',
       'start_date_format', 'end_date_format', 'Prizes', 'prize_money', 'Id',
       'Title', 'Location', 'start_date', 'end_date', 'year', 'themes',
       'prize', 'registered_N', 'featured', 'organization_name',
       'winners_announced', 'submission_gallery_url',
       'start_a_submission_url'],
      dtype='object')


Unnamed: 0                                           URL  \
0              0      https://wirvsvirushackathon.devpost.com/   
1              4            https://theglobalhack.devpost.com/   
2              8                https://supernova.devpost.com/   
3             12            https://hajjhackathon.devpost.com/   
4             16      https://chainlinkspring2022.devpost.com/   
...          ...                                           ...   
7048        7075           https://checkinproject.devpost.com/   
7049        7076                  https://fgtfrgf.devpost.com/   
7050        7077                https://old-hacks.devpost.com/   
7051        7078  https://thailand-champion-13627.devpost.com/   
7052        7079        https://manager-hackathon.devpost.com/   

                                               Criteria     schedule  \
0     Gesellschaftlicher Mehrwert, Innovationsgrad, ...  no schedule   
1     Potential impact of the project, Technical exe...  no schedule   
2     X Factor and Potential, Technical Competence, ...  no schedule   
3     Design           |             التصميم, Simpli...  no schedule   
4     User Experience / User Interface, Originality ...  no schedule   
...                                                 ...          ...   
7048                                              biden  no schedule   
7049                   Criteria 1, Criteria 1, criteria  no schedule   
7050                                                TBA  no schedule   
7051                                              biden  no schedule   
7052                                     Firts criteria  no schedule   

     hack_type                                               info  \
0       Public  Mar 29, 2020, Online, Public, Social Good     ...   
1       Public             Apr 12, 2020, Online, Public, COVID-19   
2       Public  Jul 11, 2022, Online, Public, DFINITY Foundati...   
3       Public  Aug 3, 2018, Jeddah International Exhibition &...   
4       Public  Jun 8, 2022, Online, Public, Chainlink Labs, B...   
...        ...                                                ...   
7048    Public  Aug 30, 2021, Thái Lan, Public, ACM Monterrey,...   
7049    Public  Jul 15, 2021, Online, Public, Dong Trinh LTD, ...   
7050    Public  May 7, 2021, Quận 2, Vietnam, Public, Hackers ...   
7051    Public  Sep 15, 2021, Thái Lan, Public, Screen Compose...   
7052    Public  Aug 9, 2021, Việt Nam, Public, Girls Who Code ...   

     start_date_format end_date_format  \
0           2020-03-20      2020-03-22   
1           2020-04-03      2020-04-09   
2           2022-05-10      2022-06-22   
3           2018-08-01      2018-08-03   
4           2022-04-22      2022-05-28   
...                ...             ...   
7048        2021-08-30      2021-08-30   
7049        2021-07-15      2021-07-15   
7050        2021-05-07      2021-05-07   
7051        2021-09-15      2021-09-15   
7052        2021-08-09      2021-08-09   

                                                 Prizes     prize_money  ...  \
0     Euer Preis ist Anerkennung und Ehre!, Habt Spa...  no money prize  ...   
1                                        Overall Winner  no money prize  ...   
2     SocialFi - 1ST PLACE, SocialFi - 2ND PLACE, So...        $765,000  ...   
3     1st place: 1 million Saudi Riyal (15% equity i...    SAR2,000,000  ...   
4     Chainlink Grand Prize - $40,000, Chainlink NFT...      $1,019,000  ...   
...                                                 ...             ...  ...   
7048                                         cash value          $1,000  ...   
7049               1st place, 2nd place, 3rd place, 4nd          $1,100  ...   
7050                                                TBA  no money prize  ...   
7051                                      Thailand cash            $100  ...   
7052                                                  1              $1  ...   

      end_date   year                                

In [57]:
print('Participants dataframe columns: ', participants.columns)
participants

Participants dataframe columns:  Index(['Unnamed: 0', 'name', 'skills', 'info', 'country', 'url'], dtype='object')


Unnamed: 0                               name  \
0                0                      Gerd Wittchen   
1                1                      Det. BumstΞad   
2                2                    0xNuggan Nuggan   
3                3                        Timepunk M.   
4                4                         0xaaash Li   
...            ...                                ...   
282272      282272                         Eric Zhang   
282273      282273                          Marko Zel   
282274      282274                        Mengmiao Wu   
282275      282275                         Zephyr Yao   
282276      282276  zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz   

                                                   skills  \
0                        solidity, python, react, node.js   
1                  solidity, brownie, python, react, web3   
2                                        python, solidity   
3       solidity, hardhat, thegraph, typechain, typesc...   
4                                                  python   
...                                                   ...   
282272                            javascript, python, c++   
282273                     ios, cocoapods, swift, ibeacon   
282274                                                Nan   
282275  android, python, verilog, c, c++, kernel, brow...   
282276                                      stanning-day6   

                                                     info  \
0       1 Project, 1 Hackathon, 1 Achievement, 1 Follo...   
1       1 Project, 1 Hackathon, 1 Achievement, 0 Follo...   
2       2 Projects, 3 Hackathons, 1 Achievement, 2 Fol...   
3       2 Projects, 2 Hackathons, 4 Achievements, 1 Fo...   
4       1 Project, 1 Hackathon, 1 Achievement, 0 Follo...   
...                                                   ...   
282272  1 Project, 1 Hackathon, 4 Achievements, 0 Foll...   
282273  1 Project, 1 Hackathon, 4 Achievements, 3 Foll...   
282274  1 Project, 1 Hackathon, 2 Achievements, 1 Foll...   
282275  2 Projects, 5 Hackathons, 8 Achievements, 0 Fo...   
282276  1 Project, 0 Hackathons, 2 Achievements, 0 Fol...   

                                   country                             url  
0       Sydney, New South Wales, Australia    https://devpost.com/0x1NotMe  
1                                  Francja  https://devpost.com/0xBumstead  
2                                  Germany    https://devpost.com/0xNuggan  
3                                 Portugal  https://devpost.com/0xTimepunk  
4                                       中国     https://devpost.com/0xaaash  
...                                    ...                             ...  
282272       Houston, Texas, United States                             NaN  
282273             Berlin, Berlin, Germany                             NaN  
282274            Montreal, Quebec, Canada                             NaN  
282275   Irvine, California, United States                             NaN  
282276             Ottawa, Ontario, Canada                             NaN  

[282277 rows x 6 columns]

In [58]:
print('Participants_links dataframe columns: ', participants_links.columns)
participants_links

Participants_links dataframe columns:  Index(['url', 'name', 'website', 'github', 'twitter', 'address', 'skills',
       'interests'],
      dtype='object')


url              name  \
0           https://devpost.com/CraigHiller      Craig Hiller   
1              https://devpost.com/asheleby    Andrew Sheleby   
2             https://devpost.com/evanho239           Evan Ho   
3         https://devpost.com/mag97palacios  Magnolia Quezada   
4          https://devpost.com/vivekvjk2020  Vivek Vijaykumar   
...                                     ...               ...   
191633  https://devpost.com/amulyagarimella  Amulya Garimella   
191634       https://devpost.com/hiennguyen       hien nguyen   
191635  https://devpost.com/daironrodriguez  Dairon Rodriguez   
191636         https://devpost.com/brentmin         Brent Min   
191637    https://devpost.com/JoaquinMelara    Joaquin Melara   

                                   website  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4              https://github.com/vivekvjk   
...                                    ...   
191633  https://amulyagarimella.github.io/   
191634                                 NaN   
191635  https://github.com/daironrodriguez   
191636         https://github.com/brentmin   
191637                                 NaN   

                                    github twitter          address  \
0                                      NaN     NaN              NaN   
1                                      NaN     NaN    United States   
2                                      NaN     NaN    United States   
3                                      NaN     NaN              NaN   
4              https://github.com/vivekvjk     NaN    United States   
...                                    ...     ...              ...   
191633  https://github.com/amulyagarimella     NaN    United States   
191634       https://github.com/HienNguyen     NaN   United Kingdom   
191635  https://github.com/daironrodriguez     NaN    United States   
191636         https://github.com/brentmin     NaN    United States   
191637                                 NaN     NaN               US   

                                                   skills  \
0                                                     NaN   
1              wordpress|python|tableau|excel|Blockchain|   
2       java|c++|javascript|angular.js|css|html|Blockc...   
3                                                     NaN   
4                        html|java|css|python|javascript|   
...                                                   ...   
191633  web|react|python|flask|html5|websockets|node.j...   
191634  javascript|java|machine-learning|c|python|Bloc...   
191635                               java|c++|Blockchain|   
191636  python|pandas|scikit-learn|java|javascript|htm...   
191637                                                NaN   

                                                interests  
0                                                     NaN  
1       AR/VR|Blockchain|Communication|Cybersecurity|D...  
2       Beginner Friendly|Blockchain|Cybersecurity|Des...  
3                                                     NaN  
4                                                     NaN  
...                                                   ...  
191633  Beginner Friendly|Design|Education|Health|Mach...  
191634  Blockchain|AR/VR|DevOps|Fintech|Gaming|IoT|Mac...  
191635  AR/VR|Blockchain|Cybersecurity|DevOps|IoT|Life...  
191636                               Machine Learning/AI|  
191637                                                NaN  

[191638 rows x 8 columns]

In [59]:
print('Projects dataframe columns: ', projects.columns)
projects

Projects dataframe columns:  Index(['Unnamed: 0', 'submission_gal_url', 'project_URL', 'github_links',
       'participants', 'participants_num', 'build_with', 'repo_link', 'repo',
       'submitted_to_link', 'submitted_to_name', 'submitted_to_hacks_num',
       'likes', 'comments'],
      dtype='object')


Unnamed: 0                             submission_gal_url  \
0                0  https://supernova.devpost.com/project-gallery   
1                1  https://supernova.devpost.com/project-gallery   
2                2  https://supernova.devpost.com/project-gallery   
3                3  https://supernova.devpost.com/project-gallery   
4                4  https://supernova.devpost.com/project-gallery   
...            ...                                            ...   
303675      303675    https://fgtfrgf.devpost.com/project-gallery   
303676      303676  https://old-hacks.devpost.com/project-gallery   
303677      303677  https://old-hacks.devpost.com/project-gallery   
303678      303678  https://old-hacks.devpost.com/project-gallery   
303679      303679  https://old-hacks.devpost.com/project-gallery   

                                          project_URL  \
0                https://devpost.com/software/faefolk   
1             https://devpost.com/software/hls-nzcdmt   
2       https://devpost.com/software/ic-map-collector   
3             https://devpost.com/software/kontribute   
4              https://devpost.com/software/fish-tank   
...                                               ...   
303675     https://devpost.com/software/ai-bot-avgusc   
303676       https://devpost.com/software/test-on657c   
303677    https://devpost.com/software/testing-xp1svc   
303678       https://devpost.com/software/test-on657c   
303679    https://devpost.com/software/testing-xp1svc   

                                             github_links  \
0       https://github.com/ICCards/faefolk, https://gi...   
1                             https://github.com/Hls-labs   
2                     https://github.com/stumpigit/icmaps   
3       https://github.com/teambonsai/bonsai_dapp/blob...   
4                                                     NaN   
...                                                   ...   
303675                                                NaN   
303676                                                NaN   
303677                                                NaN   
303678                                                NaN   
303679                                                NaN   

                                             participants participants_num  \
0       https://devpost.com/RAW4RMCS, https://devpost....                3   
1       https://devpost.com/Domexx, https://devpost.co...                3   
2                           https://devpost.com/stumpigit                1   
3       https://devpost.com/TsuDohNimhh, https://devpo...                2   
4                      https://devpost.com/rusty-scrivens                1   
...                                                   ...              ...   
303675                https://devpost.com/jamesnguyentech                1   
303676                      https://devpost.com/d15neylyf                1   
303677                      https://devpost.com/d15neylyf                1   
303678                      https://devpost.com/d15neylyf                1   
303679                      https://devpost.com/d15neylyf                1   

                                               build_with  \
0                       gdscript, godot, ic, motoko, rust   
1       after-effects, astro, figma, motoko, nextjs, r...   
2       motoko, ogc, openlayers, openstreetmap, svelte...   
3                                      icp, motoko, react   
4                           css, html, javascript, motoko   
...                                                   ...   
303675                                         javascript   
303676                                             heroku   
303677                                               java   
303678                                             heroku   
303679                                               java   

                                                repo_link  \
0       https://ge5fs-qyaaa-aa

### Descriptive statistics

In [60]:
# Display columns with missing values for each dataframe
print("Hackathons Missing Values:")
print(hackathons.columns[hackathons.isnull().any()])

print("\nParticipants Missing Values:")
print(participants.columns[participants.isnull().any()])

print("\nParticipants Links Missing Values:")
print(participants_links.columns[participants_links.isnull().any()])

print("\nProjects Missing Values:")
print(projects.columns[projects.isnull().any()])


Hackathons Missing Values:
Index(['Criteria', 'Prizes', 'Title', 'Location', 'start_date', 'themes',
       'organization_name'],
      dtype='object')

Participants Missing Values:
Index(['skills', 'url'], dtype='object')

Participants Links Missing Values:
Index(['website', 'github', 'twitter', 'address', 'skills', 'interests'], dtype='object')

Projects Missing Values:
Index(['github_links', 'participants', 'build_with', 'repo',
       'submitted_to_hacks_num'],
      dtype='object')


#### Count of missing values for each column in all the 4 dataframes

In [61]:
# Function to display count of missing values for each column in a given dataframe
def print_missing_values_count(df):
    missing_values_count = df.isnull().sum()
    print(missing_values_count[missing_values_count > 0])

# Display count of missing values for each column in the 'hackathons' dataframe
print("Hackathons Missing Values Count:")
print_missing_values_count(hackathons)

print("\nParticipants Missing Values Count:")
print_missing_values_count(participants)

print("\nParticipant_links Missing Values Count:")
print_missing_values_count(participants_links)

print("\nProjects Missing Values Count:")
print_missing_values_count(projects)

Hackathons Missing Values Count:
Criteria               40
Prizes                 50
Title                   1
Location              157
start_date           1128
themes               4319
organization_name    5088
dtype: int64

Participants Missing Values Count:
skills       182
url       280473
dtype: int64

Participant_links Missing Values Count:
website      108960
github       106011
twitter      179342
address       38252
skills        40265
interests     99880
dtype: int64

Projects Missing Values Count:
github_links              145720
participants                5479
build_with                 19447
repo                           2
submitted_to_hacks_num      3485
dtype: int64


In [62]:
# Function to display columns with missing values for a given dataframe
def print_columns_with_missing_values(df):
    columns_with_missing_values = df.columns[df.isnull().any()]
    print(df[columns_with_missing_values])

Prints the **all the rows** for the columns that contain **at least one missing value.**

In [63]:
# Display columns with missing values for each dataframe
print("Hackathons Rows with Missing Values:")
print_columns_with_missing_values(hackathons)

Hackathons Rows with Missing Values:
                                               Criteria  \
0     Gesellschaftlicher Mehrwert, Innovationsgrad, ...   
1     Potential impact of the project, Technical exe...   
2     X Factor and Potential, Technical Competence, ...   
3     Design           |             التصميم, Simpli...   
4     User Experience / User Interface, Originality ...   
...                                                 ...   
7048                                              biden   
7049                   Criteria 1, Criteria 1, criteria   
7050                                                TBA   
7051                                              biden   
7052                                     Firts criteria   

                                                 Prizes  \
0     Euer Preis ist Anerkennung und Ehre!, Habt Spa...   
1                                        Overall Winner   
2     SocialFi - 1ST PLACE, SocialFi - 2ND PLACE, So...   
3     1st place: 1

In [64]:
print("\nParticipants Rows with Missing Values:")
print_columns_with_missing_values(participants)


Participants Rows with Missing Values:
                                                   skills  \
0                        solidity, python, react, node.js   
1                  solidity, brownie, python, react, web3   
2                                        python, solidity   
3       solidity, hardhat, thegraph, typechain, typesc...   
4                                                  python   
...                                                   ...   
282272                            javascript, python, c++   
282273                     ios, cocoapods, swift, ibeacon   
282274                                                Nan   
282275  android, python, verilog, c, c++, kernel, brow...   
282276                                      stanning-day6   

                                   url  
0         https://devpost.com/0x1NotMe  
1       https://devpost.com/0xBumstead  
2         https://devpost.com/0xNuggan  
3       https://devpost.com/0xTimepunk  
4          https://dev

In [65]:
print("\nParticipants Rows with Missing Values:")
print_columns_with_missing_values(participants)


Participants Rows with Missing Values:
                                                   skills  \
0                        solidity, python, react, node.js   
1                  solidity, brownie, python, react, web3   
2                                        python, solidity   
3       solidity, hardhat, thegraph, typechain, typesc...   
4                                                  python   
...                                                   ...   
282272                            javascript, python, c++   
282273                     ios, cocoapods, swift, ibeacon   
282274                                                Nan   
282275  android, python, verilog, c, c++, kernel, brow...   
282276                                      stanning-day6   

                                   url  
0         https://devpost.com/0x1NotMe  
1       https://devpost.com/0xBumstead  
2         https://devpost.com/0xNuggan  
3       https://devpost.com/0xTimepunk  
4          https://dev

In [66]:
print("\nProjects Rows with Missing Values:")
print_columns_with_missing_values(projects)


Projects Rows with Missing Values:
                                             github_links  \
0       https://github.com/ICCards/faefolk, https://gi...   
1                             https://github.com/Hls-labs   
2                     https://github.com/stumpigit/icmaps   
3       https://github.com/teambonsai/bonsai_dapp/blob...   
4                                                     NaN   
...                                                   ...   
303675                                                NaN   
303676                                                NaN   
303677                                                NaN   
303678                                                NaN   
303679                                                NaN   

                                             participants  \
0       https://devpost.com/RAW4RMCS, https://devpost....   
1       https://devpost.com/Domexx, https://devpost.co...   
2                           https://devpost.com/

Now need to figure out how to handle the missing values for columns that are most relevant to examining the impact of diversity initiatives found.

### Handling missing data in hackathon df

From the hackathon df, these are the columns I think are most relevant for examining the impact of diversity initiatives: 'hack_type', 'start_date_format', 'end_date_format', 'prize_money', ‘themes', 'registered_N', 'organization_name'


In [67]:
# Check for missing values in the 'hack_type' column
print(f"Number of missing values in 'hack_type': {hackathons['hack_type'].isnull().sum()}")

# Check for missing values in the 'start_date_format' column
print(f"Number of missing values in 'start_date_format': {hackathons['start_date_format'].isnull().sum()}")

# Check for missing values in the 'end_date_format' column
print(f"Number of missing values in 'end_date_format': {hackathons['end_date_format'].isnull().sum()}")

# Check for missing values in the 'prize_money' column
print(f"Number of missing values in 'prize_money': {hackathons['prize_money'].isnull().sum()}")

# Check for missing values in the 'themes' column
print(f"Number of missing values in 'themes': {hackathons['themes'].isnull().sum()}")

# Check for missing values in the 'registered_N' column
print(f"Number of missing values in 'registered_N': {hackathons['registered_N'].isnull().sum()}")

# Check for missing values in the 'organization_name' column
print(f"Number of missing values in 'organization_name': {hackathons['organization_name'].isnull().sum()}")

Number of missing values in 'hack_type': 0
Number of missing values in 'start_date_format': 0
Number of missing values in 'end_date_format': 0
Number of missing values in 'prize_money': 0
Number of missing values in 'themes': 4319
Number of missing values in 'registered_N': 0
Number of missing values in 'organization_name': 5088


Thus, we need to handle missing values only for columns 'themes' and 'organizational_name'.
Imputing 'Not specified' in the places where missing values were previously present in the 'themes' and 'organization_name' columns. **If organization_name is important, then we can even try to mine the df (maybe using the title and URL column) to find out the organisation name.**

In [68]:
# Imputation with a Neutral Theme
hackathons['themes'].fillna(value='Not Specified', inplace=True)

# Imputation with "Not Specified"
hackathons['organization_name'].fillna(value='Not Specified', inplace=True)

hackathons[['themes','organization_name']]

themes  \
0                      Social Good, COVID-19   
1                                   COVID-19   
2        Social Good, Blockchain, Open Ended   
3                              Not Specified   
4                   Blockchain, Web, Fintech   
...                                      ...   
7048                            Voice skills   
7049  AR/VR, Machine Learning/AI, Open Ended   
7050                              Open Ended   
7051                            Voice skills   
7052                            Voice skills   

                             organization_name  
0                                Not Specified  
1                                Not Specified  
2                           DFINITY Foundation  
3                                Not Specified  
4                               Chainlink Labs  
...                                        ...  
7048                             ACM Monterrey  
7049                            Dong Trinh LTD  
7050                            Hackers Galore  
7051  Screen Composers Guild of Canada & SOCAN  
7052                        Girls Who Code UVA  

[7053 rows x 2 columns]

### Handling missing data in participants df

From the participants df, these are the columns I think are most relevant for examining the impact of diversity initiatives: ‘skills’, ‘country’


In [69]:
# Check for missing values in the 'skills' column
print(f"Number of missing values in 'skills': {participants['skills'].isnull().sum()}")

# Check for missing values in the 'country' column
print(f"Number of missing values in 'country': {participants['country'].isnull().sum()}")

Number of missing values in 'skills': 182
Number of missing values in 'country': 0


Thus, we need to handle missing values only from column 'skills'.
Given that only 182 out of 282,277 rows have missing values in the 'skills' column, it's a reasonable option to consider dropping those rows. When the number of missing values is relatively small compared to the overall size of the dataset, and it doesn't significantly impact the analysis.


In [70]:
# Drop rows with missing values in the 'skills' column
participants.dropna(subset=['skills'], inplace=True)

participants[['skills','country']]

skills  \
0                        solidity, python, react, node.js   
1                  solidity, brownie, python, react, web3   
2                                        python, solidity   
3       solidity, hardhat, thegraph, typechain, typesc...   
4                                                  python   
...                                                   ...   
282272                            javascript, python, c++   
282273                     ios, cocoapods, swift, ibeacon   
282274                                                Nan   
282275  android, python, verilog, c, c++, kernel, brow...   
282276                                      stanning-day6   

                                   country  
0       Sydney, New South Wales, Australia  
1                                  Francja  
2                                  Germany  
3                                 Portugal  
4                                       中国  
...                                    ...  
282272       Houston, Texas, United States  
282273             Berlin, Berlin, Germany  
282274            Montreal, Quebec, Canada  
282275   Irvine, California, United States  
282276             Ottawa, Ontario, Canada  

[282095 rows x 2 columns]

### Handling missing data in participants_links df

From the participants_links df, these are the columns I think are most relevant for examining the impact of diversity initiatives: ‘url’, ‘website’, ‘github’

In [71]:
# Check for missing values in the 'url' column
print(f"Number of missing values in 'url': {participants_links['url'].isnull().sum()}")

# Check for missing values in the 'website' column
print(f"Number of missing values in 'website': {participants_links['website'].isnull().sum()}")

# Check for missing values in the 'github' column
print(f"Number of missing values in 'github': {participants_links['github'].isnull().sum()}")

Number of missing values in 'url': 0
Number of missing values in 'website': 108960
Number of missing values in 'github': 106011


In [72]:
# Imputation with "Not Specified" for 'website'
participants_links['website'].fillna(value='Not Specified', inplace=True)

# Imputation with "Not Specified" for 'github'
participants_links['github'].fillna(value='Not Specified', inplace=True)

participants_links[['url','website','github']]

url  \
0           https://devpost.com/CraigHiller   
1              https://devpost.com/asheleby   
2             https://devpost.com/evanho239   
3         https://devpost.com/mag97palacios   
4          https://devpost.com/vivekvjk2020   
...                                     ...   
191633  https://devpost.com/amulyagarimella   
191634       https://devpost.com/hiennguyen   
191635  https://devpost.com/daironrodriguez   
191636         https://devpost.com/brentmin   
191637    https://devpost.com/JoaquinMelara   

                                   website                              github  
0                            Not Specified                       Not Specified  
1                            Not Specified                       Not Specified  
2                            Not Specified                       Not Specified  
3                            Not Specified                       Not Specified  
4              https://github.com/vivekvjk         https://github.com/vivekvjk  
...                                    ...                                 ...  
191633  https://amulyagarimella.github.io/  https://github.com/amulyagarimella  
191634                       Not Specified       https://github.com/HienNguyen  
191635  https://github.com/daironrodriguez  https://github.com/daironrodriguez  
191636         https://github.com/brentmin         https://github.com/brentmin  
191637                       Not Specified                       Not Specified  

[191638 rows x 3 columns]

In [73]:
# Create a binary indicator column for missing 'github'
participants_links['github_missing'] = participants_links['github'].isnull().astype(int)

### Handling missing data in projects df

From the projects df, these are the columns I think are most relevant for examining the impact of diversity initiatives: ‘github_links’, ‘participants_num’, ‘build_with’


In [74]:
# Check for missing values in the 'github_links' column
print(f"Number of missing values in 'url': {projects['github_links'].isnull().sum()}")

# Check for missing values in the 'participants' column
print(f"Number of missing values in 'participants': {projects['participants'].isnull().sum()}")

# Check for missing values in the 'participants_num' column
print(f"Number of missing values in 'participants_num': {projects['participants_num'].isnull().sum()}")

# Check for missing values in the 'build_with' column
print(f"Number of missing values in 'build_with': {projects['build_with'].isnull().sum()}")

Number of missing values in 'url': 145720
Number of missing values in 'participants': 5479
Number of missing values in 'participants_num': 0
Number of missing values in 'build_with': 19447


Thus, need to handle missing values only for columns 'url', 'participants', 'build_with'

In [75]:
# Imputation with "Not Specified" for 'github_links'
projects['github_links'].fillna(value='Not Specified', inplace=True)

# Imputation with "Not Specified" for 'build_with'
projects['build_with'].fillna(value='Not Specified', inplace=True)

In [76]:
 # Find rows with missing values in 'participants' column
rows_missing_participants = projects[projects['participants'].isnull()]

# Display the rows
rows_missing_participants

Unnamed: 0                                 submission_gal_url  \
15              15      https://supernova.devpost.com/project-gallery   
23              23      https://supernova.devpost.com/project-gallery   
39              39      https://supernova.devpost.com/project-gallery   
47              47      https://supernova.devpost.com/project-gallery   
66              66      https://supernova.devpost.com/project-gallery   
...            ...                                                ...   
303557      303557  https://eco-it-biohackathon-11439.devpost.com/...   
303609      303609    https://genhacks-iv.devpost.com/project-gallery   
303616      303616    https://genhacks-iv.devpost.com/project-gallery   
303658      303658  https://market-challenge-us.devpost.com/projec...   
303659      303659  https://market-challenge-us.devpost.com/projec...   

                                              project_URL  \
15                  https://devpost.com/software/minerium   
23                  https://devpost.com/software/openchan   
39                  https://devpost.com/software/minerium   
47                  https://devpost.com/software/openchan   
66                 https://devpost.com/software/my-status   
...                                                   ...   
303557  https://devpost.com/software/soil-health-monit...   
303609       https://devpost.com/software/toureasy-tnmwpf   
303616       https://devpost.com/software/toureasy-tnmwpf   
303658  https://devpost.com/software/michael-beirne-pr...   
303659  https://devpost.com/software/michael-beirne-pr...   

                                             github_links participants  \
15                      https://github.com/KOSASIH/Ontics          NaN   
23      https://github.com/oc-ic/openchan, https://git...          NaN   
39                      https://github.com/KOSASIH/Ontics          NaN   
47      https://github.com/oc-ic/openchan, https://git...          NaN   
66                                          Not Specified          NaN   
...                                                   ...          ...   
303557                                      Not Specified          NaN   
303609          https://github.com/beecoder-pro/Tour-easy          NaN   
303616          https://github.com/beecoder-pro/Tour-easy          NaN   
303658                                      Not Specified          NaN   
303659                                      Not Specified          NaN   

       participants_num                                         build_with  \
15                    0             css, github, html, javascript, windows   
23                    0                         javascript, motoko, svelte   
39                    0             css, github, html, javascript, windows   
47                    0                         javascript, motoko, svelte   
66                    0                                                 na   
...                 ...                                                ...   
303557                0                                            arduino   
303609                0  blockchain, bootstrap, css3, data-science-tool...   
303616                0  blockchain, bootstrap, css3, data-science-tool...   
303658                0                                         javascript   
303659                0                                         javascript   

                                                repo_link  \
15                      https://github.com/KOSASIH/Ontics   
23      https://github.com/oc-ic/openchan, https://git...   
39                      https://github.com/KOSASIH/Ontics   
47      https://github.com/oc-ic/openchan, https://git...   
66                                                    Nan   
...                                                   ...   
303557                                                Nan   
303609          https://github.com/beecoder-pro/Tour-easy   
303616          https:/

Given the observation from the output above that rows with missing values in the 'participants' column also have a corresponding value of 0 in the 'participants_num' column (which is incorrect since there cannot be 0 participants), we need to handle this inconsistency.

Imputing 'participants_num' with the Average:
This approach assumes that the average count of participants is a reasonable estimate for the missing or incorrect values. It might work well if the distribution of participants is relatively uniform across projects.

Filling 'participants' with "Not Specified":
Filling 'participants' with a placeholder like "Not Specified" is a valid choice to indicate that specific participant information is missing.

In [77]:
# Display unique values in 'participants_num' column in ascending order
unique_participants_num_sorted = sorted(projects['participants_num'].unique())
print(f"Unique values in 'participants_num' (ascending order): {unique_participants_num_sorted}")

# Display frequency of each unique value in 'participants_num' column
participants_num_frequency = projects['participants_num'].value_counts().sort_index()
print(participants_num_frequency)

Unique values in 'participants_num' (ascending order): ['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '2', '20', '21', '22', '23', '24', '25', '26', '27', '28', '3', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '4', '40', '41', '42', '43', '44', '46', '5', '51', '6', '64', '66', '7', '75', '8', '80', '9', 'Nan']
0       5479
1      96375
10       283
11       168
12       182
13        74
14       136
15        83
16        55
17        22
18        38
19        22
2      57382
20        21
21        20
22        15
23        12
24         6
25        19
26         9
27         3
28         5
3      56718
30         5
31        17
32         2
33         4
34         8
35         2
36        11
37         4
38         2
39         2
4      61963
40         2
41         3
42         4
43         2
44         4
46         2
5      14971
51         2
6       3523
64         4
66         4
7       1280
75         2
8        759
80         2
9        4

Based on the distribution of values in the 'participants_num' column, it seems that '0' and 'Nan' are outliers and make up only 3% of the dataframe, hence, we can exclude these rows

In [81]:
# Filter out rows with '0' in the 'participants_num' column
projects = projects[projects['participants_num'] != '0']

# Filter out rows with '0' in the 'participants_num' column
projects = projects[projects['participants_num'] != 'Nan']

# Display the size of the cleaned dataframe
print("Size of Cleaned DataFrame:", projects.shape)

Size of Cleaned DataFrame: (294716, 14)
